File: 0241-different-ways-to-add-parentheses/solution.cpp
vector <int> calc(vector<int> l,vector<int> r,char op)
{
    vector <int> res;
    for(auto x : l)
    {
        for(auto y : r)
        {
            if(op=='+') res.push_back(x+y);
            else if(op=='-') res.push_back(x-y);
            else res.push_back(x*y);
        }
    }
    return res;
}

vector <int> dfs(string expression)
{
    bool chk = true;
    for(auto c : expression)
    {
        if(!('0'<=c && c<='9'))
        {
            chk = false;
            break;
        }
    }
    if(chk)
    {
        return {stoi(expression)};
    }

    vector <int> res;
    int n = expression.length();
    for(int i=0;i<n;i++)
    {
        if(expression[i]=='+' || expression[i]=='-' || expression[i]=='*')
        {
            vector <int> l = dfs(expression.substr(0,i));
            vector <int> r = dfs(expression.substr(i+1));
            vector <int> result = calc(l,r,expression[i]);
            for(auto it : result) res.push_back(it); 
        }
    }
    return res;
}
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        return dfs(expression);   
    }
};

File: 0241-different-ways-to-add-parentheses/solution.py
# 괄호가 쳐져있는 식의 형태 
# (왼쪽 피연산자, 연산자, 오른쪽 피연산자)
def calc(L,R,op):
    if op == '+':
        # [1,2,3], [4,5] = [1+4, 2+4, 3+4, 1+5, 2+5, 3+5]
        return [l + r for l in L for r in R]
    elif op == '-':
        return [l - r for l in L for r in R]
    else:
        return [l * r for l in L for r in R]

def dfs(exp):
    ## 1. 숫자만 있는 경우
    ## 2. 연산자랑 같이 있는 경우

    # 1. 숫자만 있는 경우 처리
    if exp.isdigit():
        return [int(exp)]
    
    # 2. 연산자 있는 경우 처리

    # A = [1,2,3]
    # B = [4,5,6]
    # A.extend(B) -> [1,2,3,4,5,6]
    n = len(exp)
    res = []
    for i in range(n):
        if exp[i] == '+' or exp[i] == '-' or exp[i] == '*': # 현재 보고 있는 게 연산자인 경우
            L = dfs(exp[0:i])
            R = dfs(exp[i+1:])
            res.extend(calc(L,R,exp[i]))
    return res
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        return dfs(expression)        
